### ISSUE TITLE
collection_cache_key loads model when model not needed

### ISSUE THREAD

**Issue by brchristian** [rails/rails#35982]
### Steps to reproduce
**controllers/author.rb**
```ruby
def show
  @author = Author.friendly.find params[:id]
  @articles = @author.articles
end
```

**show.html.erb**
```ruby
<% cache [@author, @articles] do %>
  foo
<% end %>
```
### Expected behavior
I expect this to trigger two database queries:
(1) the model for `@author` (needed to create the ActiveRecord query for `@articles`)
(2) the cache key of `@articles`

(Note that we don’t need a query for the cache key of `@author` because the model is already loaded in memory by the time we need it.)

### Actual behavior
Actually, _three_ queries happen:
(1) the model for `@author`
(2) the cache key of `@articles`
(3) the model for `@articles` (!)

See below:

```ruby
Started GET "/authors/test-author" for ::1 at 2019-04-15 13:15:44 -0700
Processing by AuthorsController#show as HTML
  Parameters: {"id"=>"test-author"}
  Author Load (0.4ms)  SELECT  `authors`.* FROM `authors` WHERE `authors`.`slug` = ’test-author' LIMIT 1
  ↳ app/controllers/authors_controller.rb:16
  Rendering authors/show.html.erb within layouts/application
   (0.6ms)  SELECT COUNT(*) AS `size`, MAX(`articles`.`updated_at`) AS timestamp FROM `articles` INNER JOIN `authorships` ON `articles`.`id` = `authorships`.`article_id` WHERE `authorships`.`author_id` = 8
  ↳ app/views/authors/show.html.erb:3
  Article Load (0.9ms)  SELECT `articles`.* FROM `articles` INNER JOIN `authorships` ON `articles`.`id` = `authorships`.`article_id` WHERE `authorships`.`author_id` = 8 ORDER BY `articles`.`id` ASC
  ↳ app/views/authors/show.html.erb:3
Completed 200 OK in 83ms (Views: 78.4ms | ActiveRecord: 1.9ms)
```

This seems weird. Why do we need the `Article Load` query? The cache key should give us everything we need, and there is no further reference to the `@articles` object, so it should never be fully evaluated.

Am I missing something here?

### System configuration
**Rails version**: 5.2.3

**Ruby version**: 2.6.2

**Comment by st0012** [comment]
Can you create a gist/sample app that reproduces the issue? I'd be happy to help you solve this issue 😄

**Comment by brchristian** [comment]
Hi @st0012 thanks for your help!

I’m not sure how to upload a sample app, but here’s a minimal app that reproduces the problem.

```
rails new demo_app
cd demo_app
rails g scaffold author
rails g scaffold article author_id:integer
rake db:create
rake db:migrate
edit app/models/author.rb
edit app/models/article.rb
edit app/controllers/authors_controller.rb
edit app/views/authors/show.html.erb
rails dev:cache
rails s
```

**app/models/author.rb**
```ruby
class Author < ApplicationRecord
  has_many :articles
end
```

**app/models/article.rb**
```ruby
class Article < ApplicationRecord
  belongs_to :author
end
```

**app/controllers/authors_controller.rb**
```ruby
  def show
    @articles = @author.articles
  end
```

**app/views/authors/show.html.erb**
```ruby
<% cache [@author, @articles] do %>
  <%= Time.current %>
<% end %>
```

Terminal output for `http://localhost:3000/authors/1`:

```
Started GET "/authors/1" for ::1 at 2019-04-16 09:34:44 -0700
Processing by AuthorsController#show as HTML
  Parameters: {"id"=>"1"}
  Author Load (0.2ms)  SELECT  "authors".* FROM "authors" WHERE "authors"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  ↳ app/controllers/authors_controller.rb:68
  Rendering authors/show.html.erb within layouts/application
   (0.2ms)  SELECT COUNT(*) AS "size", MAX("articles"."updated_at") AS timestamp FROM "articles" WHERE "articles"."author_id" = ?  [["author_id", 1]]
  ↳ app/views/authors/show.html.erb:3
  Article Load (0.2ms)  SELECT "articles".* FROM "articles" WHERE "articles"."author_id" = ?  [["author_id", 1]]
  ↳ app/views/authors/show.html.erb:3
  Rendered authors/show.html.erb within layouts/application (14.4ms)
Completed 200 OK in 83ms (Views: 68.1ms | ActiveRecord: 0.5ms)
```

Again, note that there are **three** db queries here, when I would expect there to be two.

**Comment by brchristian** [comment]
I’ve done a bit of research and I think I have narrowed down what is happening.

The full `SELECT "articles".* FROM "articles"` query appears to be triggered by `key.to_a` in [this line](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/cache.rb#L676).

This in turn seems to have something to do with the fact that the `key` object does _not_ respond to `cache_version`.

I’m not super familiar with this part of the codebase, but something seems wrong here to me if we need to pull the entire `SELECT "articles".* FROM "articles"` into memory just to create the cache key. A big part of the point of a cache is to prevent us from grabbing unnecessary stuff out of the database! And if we _do_ need to pull the entire records, then why go through the trouble of the earlier query to grab _just_ the size and max timestamp?

**Comment by kaspth** [comment]
This is technically fixed on master with https://github.com/rails/rails/commit/1da9a7e410f15859939eb4158421846638a41e71 …if the new config is turned on.

It's definitely a bug and that line you've found is the right place to fix it. Though I'm not sure if we should just add a `is_a?(Relation)` in there. Please do investigate and PR!

**Comment by brchristian** [comment]
Hi @kaspth thanks for the helpful context!

I haven’t looked very deeply into https://github.com/rails/rails/commit/1da9a7e410f15859939eb4158421846638a41e71 or the big refactor in https://github.com/rails/rails/commit/10919bfef2303c3409b15ed1d9405f3b4c518507 yet, but trying out my demo site with the latest Rails `master`, it looks like this triple-query issue is fixed _both_ when `ActiveRecord::Base.cache_versioning = true` and when`ActiveRecord::Base.cache_versioning = false`.

LMK if you disagree, but I think my conclusion here would simply be to consider the problem fixed in 6.0, in which case we can just close this issue and thank the folks behind those other PRs!

**Comment by st0012** [comment]
Will the fix be back ported to 5.2.x?

**Comment by kaspth** [comment]
@brchristian ahh right, it's fixed all the time because Relation now responds to `cache_version`, regardless of the config, so it never hits `to_a`.

@st0012 that's a feature and not a fix so no it won't be backported. The straight fix could be to add this to `relation.rb`:

```ruby
def cache_version(timestamp_column = :updated_at) # :nodoc:
  nil
end
```

We'd need a test that asserted the right number of queries too.
